language-support/ts/codegen/src/TsCodeGenMain.hs

```diff
@@ -82,79 +111,150 @@ mergePackageMap ps = foldM merge Map.empty ps
-            ]
-        imports =
-            ["import * as " <> modNameStr <> " from '" <> pkgRootPath <> "/" <> pkgRefStr <> T.intercalate "/" (unModuleName modName) <> "';"
-            | modRef@(pkgRef, modName) <- Set.toList ((PRSelf, curModName) `Set.delete` Set.unions refs)
-            , let pkgRefStr = case pkgRef of
-                    PRSelf -> ""
-                    PRImport pkgId ->
-                      -- If the imported package has a symbolic name,
-                      -- then we need to use it in the package path.
-                      "../" <> (case Map.lookup pkgId pm of
-                                   Just (Just name, _) -> unPackageName name
-                                   Just (Nothing, _) -> unPackageId pkgId
-                                   Nothing -> error "IMPOSSIBLE : package map malformed"
-                               ) <> "/"
-            , let modNameStr = genModuleRef modRef
-            ]
-        defs = map (\(def, ser) -> def ++ ser) defSers
-    in
-    Just $ T.unlines $ intercalate [""] $ filter (not . null) $ header : imports : defs
+    let (defSers, refs) = unzip (map (genDataDef curPkgId mod tpls) serDefs)
+        imports = [ importDecl pm modRef rootPath
+                  | modRef@(pkgRef, _) <- modRefs refs
+                  , let rootPath = pkgRootPath modName pkgRef ]
+        defs = map biconcat defSers
+        modText = T.unlines $ intercalate [""] $ filter (not . null) $ modHeader : imports : defs
+        depends = [ Dependency . T.unpack $ pkgRefStr pm pkgRef
+                  | (pkgRef, _) <- modRefs refs, pkgRef /= PRSelf ]
+   in Just (modText, depends)
   where
+    modName = moduleName mod
+    tpls = moduleTemplates mod
     serDefs = defDataTypes mod
+    modRefs refs = Set.toList ((PRSelf, modName) `Set.delete` Set.unions refs)
+    modHeader =
+      [ "// Generated from " <> T.intercalate "/" (unModuleName modName) <> ".daml"
+      , "/* eslint-disable @typescript-eslint/camelcase */"
+      , "/* eslint-disable @typescript-eslint/no-use-before-define */"
+      , "import * as jtv from '@mojotech/json-type-validation';"
+      , "import * as daml from '@daml/types';"
+      ]
+
+    -- Calculate an import declaration.
+    importDecl :: Map.Map PackageId (Maybe PackageName, Package) ->
+                      (PackageRef, ModuleName) -> T.Text -> T.Text
+    importDecl pm modRef@(pkgRef, modName) rootPath =
+      "import * as " <>  genModuleRef modRef <> " from '" <>
+      T.intercalate "/" ((rootPath : pkgRefStr pm pkgRef : ["lib" | pkgRef /= PRSelf]) ++ unModuleName modName) <>
+      "';"
+
+    -- Produce a package name for a package ref.
+    pkgRefStr :: Map.Map PackageId (Maybe PackageName, Package) -> PackageRef -> T.Text
+    pkgRefStr pm = \case
+      PRSelf -> ""
+      PRImport pkgId ->
+        maybe (error "IMPOSSIBLE : package map malformed")
+        (\(mbPkgName, _) -> T.pack $ packageNameStr pkgId mbPkgName)
```

Comment:
IMO, the only point of `maybe` is to allow for point-free style:
```suggestion
        (T.pack . packageNameStr pkgId . fst)
```
I still prefer a `case ... of` expression though. It would cover the same number of lines and be easier to read.

Reply:
What you've written there is nice. Sorry I missed that. Given your comments about my use of maybes in general here I guess I'll eschew it though in favor of a case.

Reply:
✔️